addons/gamification/models/gamification_badge.py

```diff
@@ -0,0 +1,231 @@
+            else:
+                badge.remaining_sending = badge.rule_max_number - badge.stat_my_monthly_sending
+
+    def _compute_owners_info(self):
+        """Return:
+            the list of unique res.users ids having received this badge
+            the total number of time this badge was granted
+            the total number of users this badge was granted to
+        """
+        for bid in self:
+            bid.stat_count = 0
+            bid.stat_count_distinct = 0
+            bid.unique_owner_ids = []
+
+        self.env.cr.execute("""
+            SELECT badge_id, count(user_id) as stat_count,
+                count(distinct(user_id)) as stat_count_distinct,
+                array_agg(distinct(user_id)) as unique_owner_ids
+            FROM gamification_badge_user
+            WHERE badge_id in %s
+            GROUP BY badge_id
+            """, (tuple(self.ids),))
+        for (badge_id, stat_count, stat_count_distinct, unique_owner_ids) in self.env.cr.fetchall():
+            badge = self.browse(badge_id)
+            badge.stat_count = stat_count
+            badge.stat_count_distinct = stat_count_distinct
+            badge.unique_owner_ids = unique_owner_ids
+
+    def check_granting(self):
+        """Check the user 'uid' can grant the badge 'badge_id' and raise the appropriate exception
+        if not
+
+        Do not check for SUPERUSER_ID
+        """
+        status_code = self._can_grant_badge()
+        if status_code == self.CAN_GRANT:
+            return True
+        elif status_code == self.NOBODY_CAN_GRANT:
+            raise UserError(_('This badge can not be sent by users.'))
+        elif status_code == self.USER_NOT_VIP:
+            raise UserError(_('You are not in the user allowed list.'))
+        elif status_code == self.BADGE_REQUIRED:
+            raise UserError(_('You do not have the required badges.'))
+        elif status_code == self.TOO_MANY:
+            raise UserError(_('You have already sent this badge too many time this month.'))
+        else:
+            _logger.exception("Unknown badge status code: %d" % int(status_code))
+        return False
+
+    def _can_grant_badge(self):
+        """Check if a user can grant a badge to another user
+
+        :param uid: the id of the res.users trying to send the badge
+        :param badge_id: the granted badge id
+        :return: integer representing the permission.
+        """
+        if self.env.uid == SUPERUSER_ID:
+            return self.CAN_GRANT
+
+        if self.rule_auth == 'nobody':
+            return self.NOBODY_CAN_GRANT
+
+        elif self.rule_auth == 'users' and self.env.uid not in self.rule_auth_user_ids.ids:
+            return self.USER_NOT_VIP
+
+        elif self.rule_auth == 'having':
+            all_user_badges = self.env['gamification.badge.user'].search([('user_id', '=', self.env.uid)]).mapped('badge_id')
+            if set(self.rule_auth_badge_ids.ids) > set(all_user_badges.ids):
```

Comment:
What if required badges are, Good Job, Brilliant Work and All user badges are Good Skill, Bla Bla and Bla Bla Bla, should this make condition true ?